feat!: embed the transaction which spent the input DBC within the Dbc struct

bochaco opened the pull request.

bochaco committed:
feat!: embed the transaction which spent the inputs within the Dbc struct

maqi commented on src/dbc.rs:
```diff
@@ -345,8 +361,9 @@ pub(crate) mod tests {
-                dbc_builder.add_spent_proof_share(spentbook_node.log_spent(key_image, tx)?);
+            dbc_builder = dbc_builder
+                .add_spent_proof_share(spentbook_node.log_spent(key_image, tx.clone())?)
+                .add_spent_transaction(tx);
```
this looks like there will be `one transaction per proof_share` get added, which means X copies of duplicated tx get pushed to the vector?

bochaco replied:
it seems so, but with the change to using a set for the TXs then we should be fine

bochaco commented on src/builder.rs:
```diff
@@ -401,6 +407,15 @@ impl DbcBuilder {
         // note that we do this just once for entire Tx, not once per output Dbc.
         TransactionVerifier::verify(verifier, &self.transaction, &spent_proofs)?;
 
+        // verify there is a maching spent transaction for each spent_proof
```
I don't think it's necessary, what it's important is that for each proof there is a TX to verify the hash against, if there are more TXs shouldn't matter I believe?

bochaco commented on src/dbc.rs:
```diff
@@ -322,7 +338,7 @@ pub(crate) mod tests {
     }
 
-    const DBC_WITH_1_530_000_000: &str = "2fce3f70d7ad38d48d81f3f67c11ea10388a95efff66ce444f3274fb261c119a839e3de7a4c97b3b2b12fde3af09990ed20f48b3a23d79ca9b39cf0b6f492b142c03d3c0ee34d087fa3b9989cab7f7fb4e2328587330ce226277b8062881b3a231b11365712ad3a82ece13af3699fb171f16001af12522a92d1cf6d788a919bf7aa7bc1cd70f0eac4f96967bdd9cb38338aab683457882ebc8dc80cd829f9e292837a00789952636524a544ae1fada10439ed ...
+    const DBC_WITH_1_530_000_000: &str = "5b27e8998542c6ae461c20bbb764da84b16721c795fa5ec73db3d109a68dcdded655d1c1ed7d2106ac1d12558049bab64581076215747dbbff95397a32a3d3848ceb318cf6dd5b371a2e2e910e0697972fb69d93e07de0d4387c3e4dfa2d59bdf91debc36b3bc8c45c3fa390e9bbb492ba54cdaca5bd94544a56f8d209b8876fa3e5eeef1e9d624a4b65c2627983dfbc3ef0f2cb1b815c3748052525fb7bdab933a5fdfc39d7dac1f657bd63f3c64d9e7601e ...
```
correct

bochaco committed:
chore: keeping spent transactions within Dbc in a set rather than a vec

BREAKING CHANGE: changes in the API due to Dbc struct change, but also due
to moving GenesisMaterial to the mock/exmaple mod as it's not needed in public API.

maqi commented:
bors r+
